gcloud/bigquery/table.py

```diff
@@ -525,3 +551,146 @@ def delete(self, client=None):
+
+    def insert_all(self,
+                   rows,
+                   row_ids=None,
+                   skip_invalid_rows=None,
+                   ignore_unknown_values=None,
+                   client=None):
+        """API call:  insert table data via a POST request
+
+        See:
+        https://cloud.google.com/bigquery/reference/rest/v2/tabledata/insertAll
+
+        :type rows: list of tuples
+        :param rows: row data to be inserted
+
+        :type row_ids: list of string
+        :param row_ids: Unique ids, one per row being inserted.  If not
+                        passed, no de-duplication occurs.
+
+        :type skip_invalid_rows: boolean or ``NoneType``
+        :param skip_invalid_rows: skip rows w/ invalid data?
+
+        :type ignore_unknown_values: boolean or ``NoneType``
+        :param ignore_unknown_values: ignore columns beyond schema?
+
+        :type client: :class:`gcloud.bigquery.client.Client` or ``NoneType``
+        :param client: the client to use.  If not passed, falls back to the
+                       ``client`` stored on the current dataset.
+
+        :rtype: list of mappings
+        :returns: One mapping per row with insert errors:  the "index" key
+                  identifies the row, and the "errors" key contains a list
+                  of the mappings describing one or more problems with the
+                  row.
+        """
+        client = self._require_client(client)
+        rows_info = []
+        data = {'rows': rows_info}
+
+        for index, row in enumerate(rows):
+            row_info = {}
+
+            for field, value in zip(self._schema, row):
+                converter = _JSON_FROM_CELLDATA.get(field.field_type)
+                if converter is not None:
+                    value = converter(value)
+                row_info[field.name] = value
+
+            info = {'json': row_info}
+            if row_ids is not None:
+                info['insertId'] = row_ids[index]
+
+            rows_info.append(info)
+
+        if skip_invalid_rows is not None:
+            data['skipInvalidRows'] = skip_invalid_rows
+
+        if ignore_unknown_values is not None:
+            data['ignoreUnknownValues'] = ignore_unknown_values
+
+        response = client.connection.api_request(
+            method='POST',
+            path='%s/insertAll' % self.path,
+            data=data)
+        errors = []
+
+        for error in response.get('insertErrors', ()):
+            errors.append({'index': int(error['index']),
+                           'errors': error['errors']})
+
+        return errors
+
+
+def _bool_from_json(value):
+    return value.lower() in ['t', 'true', '1']
+
+
+def _datetime_from_json(value):
+    return _datetime_from_prop(float(value))
+
+
+_CELLDATA_FROM_JSON = {
+    'INTEGER': int,
+    'FLOAT': float,
+    'BOOLEAN': _bool_from_json,
+    'TIMESTAMP': _datetime_from_json,
+}
+
+_JSON_FROM_CELLDATA = {
+    'TIMESTAMP': _prop_from_datetime,
+}
```

Comment:
> Can you put these at the top of the file?

Why  care?  They are private implementaition details.

Enums still feel like YAGNI to me (unless we decide to move to using them across all APIs).